vmdb/app/models/mixins/process_tasks_mixin.rb

```diff
@@ -0,0 +1,123 @@
+module ProcessTasksMixin
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    # Processes tasks received from the UI and queues them
+    def process_tasks(options)
+      raise "No ids given to process_tasks" if options[:ids].blank?
+      if options[:task] == "refresh_ems" && respond_to?("refresh_ems")
+        refresh_ems(options[:ids])
+        msg = "'#{options[:task]}' initiated for "\
+              "#{options[:ids].length} #{ui_lookup(:table => base_class.name).pluralize}"
+        task_audit_event(:success, options, :message => msg)
+      else
+        exception_if_task_unknown(options)
+        options[:userid] ||= "system"
+        invoke_tasks_queue(options)
+      end
+    end
+
+    def invoke_tasks_queue(options)
+      MiqQueue.put(:class_name => name, :method_name => "invoke_tasks", :args => [options])
+    end
+
+    # Performs tasks received from the UI via the queue
+    def invoke_tasks(options)
+      local, remote = partition_ids_by_remote_region(options[:ids])
+      invoke_tasks_local(options.merge(:ids => local)) unless local.empty?
+
+      # TODO: invoke_tasks_remote currently is only implemented by VmOrTemplate.
+      # it can be refactored to be generalized like invoke_tasks_local
+      invoke_tasks_remote(options.merge(:ids => remote)) if !remote.empty? && respond_to?("invoke_tasks_remote")
+    end
+
+    def invoke_tasks_local(options)
+      options[:invoke_by] = task_invoked_by(options)
+      args = task_arguments(options)
+
+      instances, tasks = validate_tasks(options)
+
+      instances.each_with_index do |instance, idx|
+        task = MiqTask.where(:id => tasks[idx]).first
```

Comment:
You can avoid this N+1 done by this loop.

- First, to make it clearer, rename `tasks` to `task_ids`.
- Query all the tasks up front outside the loop, and index them by task.id: `tasks_by_id = MiqTasks.where(:id => task_ids).index_by(:id)`
- No need for the index, use `.zip` instead

```ruby
instances, task_ids = validate_tasks(options)
tasks_by_id = MiqTasks.where(:id => task_ids).index_by(&:id)
tasks = task_ids.collect { |task_id| tasks[task_id] }

instances.zip(tasks).each do |instance, task|
  ...
```